src/types/int_type.rs

```diff
@@ -209,29 +237,27 @@ impl IntType {
     ///
     /// ```no_run
     /// use inkwell::context::Context;
+    /// use inkwell::types::StringRadix;
     ///
     /// let context = Context::create();
     /// let i8_type = context.i8_type();
-    /// let i8_val = i8_type.const_int_from_string("0121", 10);
+    /// let i8_val = i8_type.const_int_from_string("0121", StringRadix::Decimal);
     ///
     /// assert_eq!(i8_val.print_to_string().to_string(), "i8 121");
     ///
-    /// let i8_val = i8_type.const_int_from_string("0121", 3);
+    /// let i8_val = i8_type.const_int_from_string("0121", StringRadix::from_u8(10).unwrap());
     ///
     /// assert_eq!(i8_val.print_to_string().to_string(), "i8 16");
     ///
-    /// // Unexpected outputs:
-    /// let i8_val = i8_type.const_int_from_string("0121", 2);
-    ///
-    /// assert_eq!(i8_val.print_to_string().to_string(), "i8 3");
-    ///
-    /// let i8_val = i8_type.const_int_from_string("ABCD", 2);
+    /// // Unsafe executions, LLVM may terminate with an assertion failure or
+    /// // execute undefined behaviour in C++ code.
```

Comment:
I'm not sure I follow as to how this can be UB? If so we should make the function unsafe?

Reply:
LLVM's API is pretty crazy -- the caller is responsible for making sure that all inputs follow all of llvm's rules (and the documentation is not always there). If you don't, LLVM may do absolutely anything including UB. So, from llvm's point of view, asking it to parse "3.asd" in base 10, llvm is permitted to perform UB since you shouldn't be asking it that. (Their point of view is that their job is to provide maximum performance for customers who want it, and even the act of checking for validity takes time which is an expense upon those who don't need it.)

As for how this particular function call can trigger UB, here's a snippet of C++ code in APInt::fromString:
```
 void APInt::fromString(unsigned numbits, StringRef str, uint8_t radix) {
   // Check our assumptions here
   assert(!str.empty() && "Invalid string length");
   assert((radix == 10 || radix == 8 || radix == 16 || radix == 2 ||
           radix == 36) &&
          "Radix should be 2, 8, 10, 16, or 36!");
 
   StringRef::iterator p = str.begin();
   size_t slen = str.size();
   bool isNeg = *p == '-';
```
StringRef is not NUL terminated, so an empty string means there are zero bytes that are safe to dereference. In an asserts build of LLVM, it will assert on an empty string, but on a non-asserts build it will dereference a pointer to no particular place in `bool isNeg = *p == '-';`.

Should we make it unsafe? I was wondering what your goal for inkwell is. Given llvm's API design, it seems like nearly every function would end up being unsafe? I'd be interested in trying to figure out how to make them safe?

Reply:
Well the function being safe is a matter of maintaining LLVM's invariants. So I think the best thing to do is either A) change the return type to be a `Result` so that the function is fallible (very verbose) or B) just assert that the string length is non zero, and document that the function call can panic if an empty string. In practice this shouldn't come up anyway, so I think this is the better of the two

Reply:
It's not just the empty string, I think the best way is to match it against a regex verifying that only the correct digits are supplied for the given radix. A panic!() is fine with me.

Reply:
Sounds good

Reply:
Done.